go/vt/discovery/healthcheck.go

```diff
@@ -349,106 +349,74 @@ func (hc *HealthCheckImpl) checkConn(hcc *healthCheckConn, name string) {
 	}
 	hc.initialUpdatesWG.Done()
 
-	// retry health check if it fails
+	// Read stream health responses.
 	for {
-		// Try to connect to the tablet.
-		stream, err := hcc.connect(hc)
+		err := hcc.stream(hc, func(shr *querypb.StreamHealthResponse) error {
+			return hcc.processResponse(hc, shr)
+		})
 		if err != nil {
-			select {
-			case <-hcc.ctx.Done():
-				return
-			default:
-			}
 			hcc.mu.Lock()
-			hcc.tabletStats.Serving = false
-			hcc.tabletStats.LastError = err
-			target := hcc.tabletStats.Target
+			ts := hcc.tabletStats
 			hcc.mu.Unlock()
-			hcErrorCounters.Add([]string{target.Keyspace, target.Shard, topoproto.TabletTypeLString(target.TabletType)}, 1)
-
-			// Sleep until the next retry is up or the context is done/canceled.
-			select {
-			case <-hcc.ctx.Done():
-			case <-time.After(hc.retryDelay):
+			// notify downstream for serving status change
+			if hc.listener != nil {
+				hc.listener.StatsUpdate(&ts)
 			}
-			continue
+			hcErrorCounters.Add([]string{ts.Target.Keyspace, ts.Target.Shard, topoproto.TabletTypeLString(ts.Target.TabletType)}, 1)
 		}
 
-		// Read stream health responses.
-		for {
-			reconnect, err := hcc.processResponse(hc, stream)
-			if err != nil {
-				hcc.mu.Lock()
-				hcc.tabletStats.Serving = false
-				hcc.tabletStats.LastError = err
-				ts := hcc.tabletStats
-				hcc.mu.Unlock()
-				// notify downstream for serving status change
-				if hc.listener != nil {
-					hc.listener.StatsUpdate(&ts)
-				}
-				select {
-				case <-hcc.ctx.Done():
-					return
-				default:
-				}
-				hcErrorCounters.Add([]string{ts.Target.Keyspace, ts.Target.Shard, topoproto.TabletTypeLString(ts.Target.TabletType)}, 1)
-				if reconnect {
-					hcc.mu.Lock()
-					hcc.conn.Close(hcc.ctx)
-					hcc.conn = nil
-					hcc.tabletStats.Target = &querypb.Target{}
-					hcc.mu.Unlock()
-
-					// Sleep until the next retry is up or the context is done/canceled.
-					select {
-					case <-hcc.ctx.Done():
-					case <-time.After(hc.retryDelay):
-					}
-					break
-				}
-			}
+		// Sleep until the next retry is up or the context is done/canceled.
+		select {
+		case <-hcc.ctx.Done():
+			return
+		case <-time.After(hc.retryDelay):
 		}
 	}
 }
 
 // connect creates connection to the tablet and starts streaming.
-func (hcc *healthCheckConn) connect(hc *HealthCheckImpl) (tabletconn.StreamHealthReader, error) {
-	// Keyspace, shard and tabletType are the ones from the tablet
-	// record, but they won't be used just yet.
-	conn, err := tabletconn.GetDialer()(hcc.tabletStats.Tablet, hc.connTimeout)
-	if err != nil {
-		return nil, err
-	}
-	stream, err := conn.StreamHealth(hcc.ctx)
-	if err != nil {
-		conn.Close(hcc.ctx)
-		return nil, err
-	}
+func (hcc *healthCheckConn) stream(hc *HealthCheckImpl, sendReply func(*querypb.StreamHealthResponse) error) error {
```

Comment:
`reply` or `response` can be confused as var names.
`stream` is already taken by gRPC.
`process` is too generic.
We could go with something neutral like `callback`.
Or, we could stay with `sendReply` until we think of a better name. It may be the least confusing.